### PR TITLE
Feature/tsg

### PR DESCRIPTION
**Pull request by hroest**
Work on TheoreticalSpectrumGenerator

- allow choice of isotopic generator (coarse / fine)
- speedup if no sorting is needed
- general code improvements
- substantial speedup for loss calculation (fast path)

# Benchmarking

Benchmark on peptide `DFPIANGER`, all measurements are against Release 2.4 (pyopenms). Benchmark code:

```  
  auto tsg= TheoreticalSpectrumGenerator();
  auto pep = AASequence::fromString("DFPIANGER");
  auto spec = MSSpectrum();
  
  for (Size i = 0; i < nr_mol;  i++) 
  {
    tsg.getSpectrum(spec, pep, 1,2);
    nr_fragments += spec.size();
    spec.clear(true);
  }
```

where `false` was used in the last line if no annotations are present.

## Single y ion series calculation

before: 1.2 usec / loop
after: 0.25 usec / loop (no sorting)
after: 0.71 usec / loop (with sorting)

single ion series calculations are ca 5x faster which is due code improvements, turning off sorting and to removing the `shrink_to_fit` from the `clear()`. Pure code improvements in TSG alone are about 50% faster (see below).

### Single y ion series calculation (TSG only)

Using `spec.clear(true);` forcing the library to clear all arrays and `shrink_to_fit` all arrays and turning on sorting, we get

before: 1.5 usec / loop
after: 0.97 usec / loop

## Single y ion series calculation with losses

before: 429 usec / loop
after: 4.9 usec / loop (no sorting)

This is mainly due to calls of `EmpiricalFormula::getMonoWeight()` since the losses are coded to individual residues and the weight of the loss needs to be re-computed each time.  This is still substantially faster than before.

## Single y ion series calculation with losses and annotation

before: 528 usec / loop
after: 34 usec / loop (no sorting)

This is mainly due to calls of `EmpiricalFormula::toString()` which is pretty slow as well as storing the annotations and dealing with Strings in general.

## Single y ion series calculation with losses, annotation and isotopes (coarse)

before: 1040 usec / loop
after: 640 usec / loop (no sorting)

## Single y ion series calculation with losses, annotation and isotopes (fine)

before: NA 
after: 1560 usec / loop (no sorting)

### CODE REVIEW

File: src/tests/class_tests/openms/source/CoarseIsotopeDistribution_test.cpp [hunk 1]
```diff
@@ -143,6 +143,39 @@ START_SECTION(IsotopeDistribution convolve_(const CoarseIsotopePatternGenerator&
...
+
+    TEST_REAL_SIMILAR(id[0].getMZ(), 180.063)
+    TEST_REAL_SIMILAR(id[0].getIntensity(), 0.923456)
+
+    TEST_REAL_SIMILAR(id[2].getMZ(), 182.0701)
+    TEST_REAL_SIMILAR(id[2].getIntensity(), 0.013232)
+  }
+
+  // TODO: is that a good idea?
+  ef.setCharge(2);
```
Comment (hroest) [inline src/tests/class_tests/openms/source/CoarseIsotopeDistribution_test.cpp]: during testing I discovered that the `CoarseIsotopePatternGenerator` does not actually compute the isotopic envelope of the given sum formula but it uses `SumFormula + ch * H` if charge is set. This is i) unexpected and ii) assumes that the adduct is hydrogen. We should probably change that.

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp [hunk 2]
```diff
@@ -359,19 +455,48 @@ END_SECTION
 
 START_SECTION(([EXTRA] bugfix test where losses lead to formulae with negative element frequencies))
 {
+  // this tests for the loss of CONH2 on Arginine, however it is not clear how
+  // this loss would occur in the first place.
   AASequence tmp_aa = AASequence::fromString("RDAGGPALKK");
```
Comment (hroest) [inline src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp]: this does not make sense to me, it was introduced in 74e2ce6761e4a273164b29b8be487 and maybe @cbielow has some insight, but to me it seems like its an error in Residues.xml -- see #4448

File: src/openms/source/KERNEL/MSSpectrum.cpp [hunk 3]
```diff
@@ -166,10 +166,11 @@ namespace OpenMS
   void MSSpectrum::clear(bool clear_meta_data)
   {
     ContainerType::clear();
-    ContainerType::shrink_to_fit();
```
Comment (hroest) [inline src/openms/source/KERNEL/MSSpectrum.cpp]: @jpfeuffer this comes from the proteomicLFQ branch de7d4d019551dc3f5972d1c353ddd5fbd and at least here (for benchmark workloads) it does make a substantial difference of about 50% in speed due to re-allocation of memory. In general, I think its confusing if `clear()` does other things in OpenMS than what one is used from the STL.

Comment (timosachsenberg) [reply src/openms/source/KERNEL/MSSpectrum.cpp]: I think we can change that back and just expose shrink to fit externally.

**Comment by timosachsenberg** [comment]
Some nice additions and optimizations. Maybe @enetz can also provide some insights on e.g. sorting .

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp (hunk 2 above)
Comment (hroest) [reply src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp]: since the losses we look at here are amino-acid specific losses, these should all be part of the side chain and not the peptide back bone

**Comment by hroest** [comment]
@enetz I think there may be an easy way to add a simple isotopic model to the TSG, basically one that does not really take intensity into account but simply adds a peak at a given intensity and C12-C13 carbon mass away. I think doing that could actually reduce the execution time from ms to usec, speeding up TSG at least 100 or even more fold. What do you think, would that be interesting?

**Comment by enetz** [comment]
@hroest in my very specialized TSG for XL-MS I already do that. Because I deisotope anything with 3 or more peaks in the experimental spectra anyway, I just add a second peak for each fragment like this:
```
        spectrum.emplace_back(pos, charge);
        if (add_isotopes_)
        {
          spectrum.emplace_back(pos+(Constants::C13C12_MASSDIFF_U / charge), charge);
        }
```
This could also be generalized for higher numbers of peaks with a simple loop. I think most of our tools do not use the intensity of theoretical peaks anyway.

**Comment by enetz** [comment]
About sorting, in my tool the sorting of theoretical spectra is still a big chunk of total runtime.
I tested several sorting algorithms and boost::pdqsort turned out to be the most efficient for the two types of spectra that OpenPepXL generates. E.g. creating 2 million linear ion spectra (the type most similar to normal peptide spectra) took 23s with std::sort and 18s with boost::pdqsort.  Also to what extent the spectra are already partially sorted can have an influence on sorting time as well, so if possible it helps to add peaks in order as much as possible, e.g. first add neutral loss peaks, then monoisotopic, then additional isotopic pattern peaks.

File: src/tests/class_tests/openms/source/CoarseIsotopeDistribution_test.cpp (hunk 1 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/CoarseIsotopeDistribution_test.cpp]: I agree and close this subconversation because we have an issue for that

**Comment by timosachsenberg** [comment]
ok I think this is good to merge as is... any more comments on the neutral loss? my take is to remove it if we can't find out the reason why its there

**Comment by cbielow** [comment]
>Also to what extent the spectra are already partially sorted can have an influence on sorting time as 
>well, so if possible it helps to add peaks in order as much as possible, e.g. first add neutral loss 
>peaks, then monoisotopic, then additional isotopic pattern peaks.

have you tried storing the each type of loss in separate vectors and use https://en.cppreference.com/w/cpp/algorithm/merge on the result? (C17 also has https://en.cppreference.com/w/cpp/algorithm/inplace_merge but alas...)
